src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,348 @@
+            this.WriteSeparatorIfNecessary();
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.binaryValueStream = new ODataUtf8JsonWriteStream(this);
+
+            return this.binaryValueStream;
+        }
+
+        /// <summary>
+        /// Asynchronously ends a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndStreamValueScopeAsync()
+        {
+            await this.binaryValueStream.DisposeAsync();
+            this.binaryValueStream = null;
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Represents a stream for writing UTF-8 JSON data to an ODataUtf8JsonWriter.
+        /// </summary>
+        private class ODataUtf8JsonWriteStream : Stream
+        {
+            private readonly ODataUtf8JsonWriter jsonWriter = null;
+            private byte[] buffer;
+            int numBytesNotWrittenFromPreviousChunk = 0;
+
+            /// <summary>
+            /// Initializes a new instance of the ODataUtf8JsonWriterStream class with the specified ODataUtf8JsonWriter.
+            /// </summary>
+            /// <param name="writer">The OData UTF-8 JSON writer to write to.</param>
+            internal ODataUtf8JsonWriteStream(ODataUtf8JsonWriter writer)
+            {
+                this.jsonWriter = writer;
+            }
+
+            public override bool CanRead => false;
+
+            public override bool CanSeek => false;
+
+            public override bool CanWrite => true;
+
+            /// <summary>
+            /// Gets the length in bytes of the stream. This is not supported by this stream.
+            public override long Length => throw new NotSupportedException();
+
+            /// <summary>
+            /// Gets or sets the position within the stream. This is not supported by this stream.
+            /// </summary>
+            public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                if (numBytesNotWrittenFromPreviousChunk > 0)
+                {
+                    // If there are unprocessed bytes, encode and write them as the final block.
+                    ReadOnlySpan<byte> bytesNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(0, numBytesNotWrittenFromPreviousChunk);
+
+                    this.jsonWriter.Base64EncodeAndWriteChunk(bytesNotProcessedFromPreviousChunk, isFinalBlock: true, out numBytesNotWrittenFromPreviousChunk);

```

Comment:
I'm concerned about setting the `isFinalBlock` to true here. It's possible that the stream is flushed before we reached the end of the of contents. The called might flush multiple times in the course of writing. I think we should only set `isFinalBlock` to true when the stream is disposed. We should have tests for this as well (i.e. should have test that should cause the code as-is to fail, a test that would write unexpected output if the flush sets isFinalBlock too prematurely).

Reply:
Maybe we should have an private `Flush(bool isFinalBlock)` helper method. Then in the `Flush()` method we would call `Flush(isFinalBlock: false)` and in `Dipose` we call `Flush(isFinalBlock: true)`. When `isFinalBlock` is true, we would also free the buffer.

Reply:
I think this logic then can be moved to the `Dispose` method.. If they flush and continue writing then in the next write..the unprocessed chars will still be there and will be processed in the next write call. The idea of including this logic here was to ensure that at the end of writing (that's when we dispose or when we done writing everything) we do not have any unprocessed bytes.